group-coordinator/src/main/java/org/apache/kafka/coordinator/group/assignor/UniformHeterogeneousAssignmentBuilder.java

```diff
@@ -157,716 +240,779 @@ public GroupAssignment build() {
-            // Remove the member's assignment data from the set to update it.
-            sortedMembersByAssignmentSize.remove(memberId);
-            assignmentManager.decrementTargetAssignmentSize(memberId);
+        memberTargetAssignmentSizes[memberIndex]++;
+    }
 
-            // Update current assignment size and re-add to set if needed.
-            if (!isMemberAtMaxCapacity(memberId)) {
-                sortedMembersByAssignmentSize.add(memberId);
+    /**
+     * Revokes the partition from a member and updates the current target assignment size.
+     *
+     * @param topicId               The topic containing the partition to be revoked.
+     * @param partition             The partition to be revoked.
+     * @param memberIndex           Member that the partition needs to be revoked from.
+     */
+    private void removePartitionFromTargetAssignment(
+        Uuid topicId,
+        int partition,
+        int memberIndex
+    ) {
+        String memberId = memberIds.get(memberIndex);
+        Map<Uuid, Set<Integer>> assignment = targetAssignment.get(memberId).partitions();
+        if (isImmutableMap(assignment)) {
+            assignment = deepCopy(assignment);
+            targetAssignment.put(memberId, new MemberAssignmentImpl(assignment));
+        }
+        Set<Integer> partitionsSet = assignment.get(topicId);
+        // Remove the partition from the assignment, if there are no more partitions from a particular topic,
+        // remove the topic from the assignment as well.
+        if (partitionsSet != null) {
+            partitionsSet.remove(partition);
+            if (partitionsSet.isEmpty()) {
+                assignment.remove(topicId);
             }
         }
 
-        /**
-         * Sorts members in ascending order based on their current target assignment size.
-         * Members that have reached their max assignment size are removed.
-         *
-         * @param memberIds     Member Ids that need to be sorted.
-         * @return A set that maintains the order of members by assignment size.
-         */
-        private TreeSet<String> sortMembersByAssignmentSize(Collection<String> memberIds) {
-            Comparator<String> comparator = Comparator
-                .comparingInt((String memberId) -> membersWithAssignmentSizes.get(memberId).currentAssignmentSize)
-                .thenComparing(memberId -> memberId);
-
-            return memberIds.stream()
-                .filter(memberId -> {
-                    MemberAssignmentData memberData = membersWithAssignmentSizes.get(memberId);
-                    return memberData.currentAssignmentSize < memberData.maxAssignmentSize;
-                })
-                .collect(Collectors.toCollection(() -> new TreeSet<>(comparator)));
+        targetAssignmentPartitionOwners.get(topicId)[partition] = -1;
+
+        memberTargetAssignmentSizes[memberIndex]--;
+    }
+
+    private static Map<Uuid, Set<Integer>> deepCopy(Map<Uuid, Set<Integer>> map) {
```

Comment:
We have the very same method in UniformHomogeneousAssignmentBuilder. I wonder if we could share them. We could also potentially share isImmutableMap too. One way would be to have a small class of helpers. Have you considered it?

Reply:
I've factored these out into an `AssignorHelpers` class.